hu.elte.refjava.lang/src/hu/elte/refjava/api/patterns/ASTBuilder.xtend

```diff
@@ -1,58 +1,310 @@
+import java.lang.reflect.Type
 import java.util.List
 import java.util.Map
+import java.util.Queue
 import org.eclipse.jdt.core.dom.AST
 import org.eclipse.jdt.core.dom.ASTNode
+import org.eclipse.jdt.core.dom.Expression
+import org.eclipse.jdt.core.dom.MethodDeclaration
+import org.eclipse.jdt.core.dom.Modifier.ModifierKeyword
 import org.eclipse.jdt.core.dom.rewrite.ASTRewrite
 
 class ASTBuilder {
 
-	val Pattern pattern
-
+	Pattern pattern
 	AST ast
 	ASTRewrite rewrite
 	Map<String, List<? extends ASTNode>> bindings
-
+	Map<String, String> nameBindings
+	Map<String, Type> typeBindings
+	Map<String, List<Pair<Type, String>>> parameterBindings
+	Queue<String> typeReferenceQueue
+	
 	new(Pattern pattern) {
 		this.pattern = pattern
 	}
 
 	def getRewrite() {
 		rewrite
 	}
+	
+	def build(Pattern pattern, AST ast, Map<String, List<? extends ASTNode>> bindings, Map<String, String> nameBindings, Map<String, Type> typeBindings, Map<String, List<Pair<Type, String>>> parameterBindings, String typeRefString) {
+		this.pattern = pattern
+		build(ast, bindings, nameBindings, typeBindings, parameterBindings, typeRefString)
+	}
+	
+	def buildNewInterface(PLambdaExpression lambdaExpr, AST ast, Map<String, List<? extends ASTNode>> bindings, Map<String, String> nameBindings, Map<String, Type> typeBindings, Map<String, List<Pair<Type, String>>> parameterBindings, String typeRefString){
+		this.ast = ast
+		
+		val newInterface = ast.newTypeDeclaration
+		newInterface.interface = true
+		newInterface.name.identifier = "test"
+		
+		if(typeRefString !== null) {
+			val tmp = typeRefString.split("\\|")
+			this.typeReferenceQueue = newLinkedList
+			this.typeReferenceQueue.addAll(tmp)
+		}
+		
+		val lambdaExpressionBody = ((lambdaExpr.expression as PMemberFeatureCall).memberCallTarget as PConstructorCall).elements
+		val newInterfaceBodyDeclarations = lambdaExpressionBody.doBuildPatterns
+		
+		newInterfaceBodyDeclarations.forEach[
+			if(it instanceof MethodDeclaration) {
+				(it as MethodDeclaration).body = null
+			}
+		]
+		
+		newInterface.bodyDeclarations.addAll(newInterfaceBodyDeclarations)
+		return newInterface
+	}
 
-	def build(AST ast, Map<String, List<? extends ASTNode>> bindings) {
+	def build(AST ast, Map<String, List<? extends ASTNode>> bindings, Map<String, String> nameBindings, Map<String, Type> typeBindings, Map<String, List<Pair<Type, String>>> parameterBindings, String typeRefString) {
 		this.ast = ast
 		this.rewrite = ASTRewrite.create(ast)
 		this.bindings = bindings
-
-		return pattern.patterns.doBuildPatterns
+		this.nameBindings = nameBindings
+		this.typeBindings = typeBindings
+		this.parameterBindings = parameterBindings
+		
+		if(typeRefString !== null) {
+			val tmp = typeRefString.split("\\|")
+			this.typeReferenceQueue = newLinkedList
+			this.typeReferenceQueue.addAll(tmp)
+		}
+		
+		if(pattern.empty) {
+			return null
+		} else {
+			return pattern.patterns.doBuildPatterns
+		}
+		
 	}
 
+	//meta variable builder
```

Comment:
Yes it is, but I wanted to do comments on all of that, because my mind was blown at some point. I will do a proper commenting when everything is done.